src/resolve/resolved-manager.c

```diff
@@ -1054,6 +1054,65 @@ static int manager_ipv6_send(
         return sendmsg_loop(fd, &mh, 0);
 }
 
+int send_dns_notification(Manager *m, DnsAnswer *answer, const char *query_name) {
+        _cleanup_free_ char *normalized = NULL;
+        DnsResourceRecord *rr;
+        int ifindex, r;
+        _cleanup_(json_variant_unrefp) JsonVariant *array = NULL;
+        Varlink *connection;
+
+        assert(m);
+
+        if (set_isempty(m->varlink_subscription))
+                return 0;
+
+        DNS_ANSWER_FOREACH_IFINDEX(rr, ifindex, answer) {
+                _cleanup_(json_variant_unrefp) JsonVariant *entry = NULL;
+
+                if (rr->key->type == DNS_TYPE_A) {
+                        struct in_addr *addr = &rr->a.in_addr;
+                        r = json_build(&entry,
+                                JSON_BUILD_OBJECT(JSON_BUILD_PAIR_CONDITION(ifindex > 0, "ifindex", JSON_BUILD_INTEGER(ifindex)),
+                                                  JSON_BUILD_PAIR_INTEGER("family", AF_INET),
+                                                  JSON_BUILD_PAIR_IN4_ADDR("address", addr),
+                                                  JSON_BUILD_PAIR_STRING("type", "A")));
+                } else if (rr->key->type == DNS_TYPE_AAAA) {
+                        struct in6_addr *addr6 = &rr->aaaa.in6_addr;
+                        r = json_build(&entry,
+                                JSON_BUILD_OBJECT(JSON_BUILD_PAIR_CONDITION(ifindex > 0, "ifindex", JSON_BUILD_INTEGER(ifindex)),
+                                                  JSON_BUILD_PAIR_INTEGER("family", AF_INET6),
+                                                  JSON_BUILD_PAIR_IN6_ADDR("address", addr6),
+                                                  JSON_BUILD_PAIR_STRING("type", "AAAA")));
+                } else
```

Comment:
I am not convinced this is the way to do it. Either pass along the binary record data, encoding in base64, or adds a function resource_record_to_json() that add mappings for *all* record types we can currently parse.